File: src/lib/mechanics/modifiers.ts
// Modifier Types - Extended for enhanced mechanics
export type ModifierType =
    | 'bonus'
    | 'set'
    | 'override'
    | 'proficiency'
    | 'language'
    | 'expertise'
    | 'skill_proficiency'
    // New modifier types
    | 'ability_increase'
    | 'ability_point_grant'  // Grants points to ability point pool (e.g., ASI feat)
    | 'saving_throw_proficiency'
    | 'armor_proficiency'
    | 'weapon_proficiency';

export interface Modifier {
    type: ModifierType;
    target: string;
    value: number | boolean | string;
    condition?: string;
    max?: number; // Maximum cap for ability_increase modifiers (e.g., 20)
}

export interface ModifierSource {
    id: string;
    name: string;
    modifiers: Modifier[];
}

// Constants
export const MAX_ATTUNED_ITEMS = 3;

/**
 * Calculates the final value of a stat by applying all modifiers.
 * Order of operations:
 * 1. Base value
 * 2. 'set' modifiers (highest wins)
 * 3. 'bonus' and 'ability_increase' modifiers (additive)
 * 4. Apply max caps from modifiers
 */
export function calculateStatWithModifiers(baseValue: number, statName: string, sources: ModifierSource[]): number {
    let finalValue = baseValue;
    const bonuses: number[] = [];
    let overrideValue: number | null = null;
    let minMaxCap: number | null = null; // Track the lowest max cap from all modifiers

    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.target.toLowerCase() !== statName.toLowerCase()) continue;

            // TODO: Handle conditions (e.g., "while wearing heavy armor")
            if (mod.condition) continue;

            if (mod.type === 'set' || mod.type === 'override') {
                const val = typeof mod.value === 'number' ? mod.value : parseInt(String(mod.value), 10);
                if (!isNaN(val)) {
                    // "Set" means "At least this much" - take highest override
                    if (overrideValue === null || val > overrideValue) {
                        overrideValue = val;
                    }
                }
            } else if (mod.type === 'bonus' || mod.type === 'ability_increase') {
                const val = typeof mod.value === 'number' ? mod.value : parseInt(String(mod.value), 10);
                if (!isNaN(val)) {
                    bonuses.push(val);
                }

                // Track max cap from ability_increase modifiers
                if (mod.type === 'ability_increase' && mod.max !== undefined) {
                    if (minMaxCap === null || mod.max < minMaxCap) {
                        minMaxCap = mod.max;
                    }
                }
            }
        }
    }

    // Apply override if present
    if (overrideValue !== null) {
        finalValue = overrideValue;
    }

    // Apply bonuses
    finalValue = finalValue + bonuses.reduce((a, b) => a + b, 0);

    // Apply max cap if present
    if (minMaxCap !== null) {
        finalValue = Math.min(finalValue, minMaxCap);
    }

    return finalValue;
}

/**
 * Checks if a skill/stat has proficiency granted via modifiers.
 */
export function getProficiencyModifiers(statName: string, sources: ModifierSource[]): boolean {
    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'skill_proficiency' && mod.target.toLowerCase() === statName.toLowerCase() && mod.value === true) {
                return true;
            }
            if (mod.type === 'proficiency' && mod.target.toLowerCase() === statName.toLowerCase()) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Checks if character has saving throw proficiency via modifiers.
 */
export function getSavingThrowProficiency(ability: string, sources: ModifierSource[]): boolean {
    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'saving_throw_proficiency' &&
                mod.target.toLowerCase() === ability.toLowerCase() &&
                mod.value === true) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Checks if character has armor proficiency via modifiers.
 * @param armorType - 'light', 'medium', 'heavy', or 'shields'
 */
export function getArmorProficiency(armorType: string, sources: ModifierSource[]): boolean {
    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'armor_proficiency' &&
                mod.target.toLowerCase() === armorType.toLowerCase() &&
                mod.value === true) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Checks if character has weapon proficiency via modifiers.
 * @param weaponIdOrCategory - A weapon ID (e.g., 'longsword') or category ('simple', 'martial')
 */
export function getWeaponProficiency(weaponIdOrCategory: string, sources: ModifierSource[]): boolean {
    const target = weaponIdOrCategory.toLowerCase();

    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'weapon_proficiency' && mod.value === true) {
                const modTarget = mod.target.toLowerCase();
                // Direct match (specific weapon or category)
                if (modTarget === target) {
                    return true;
                }
            }
        }
    }
    return false;
}

/**
 * Gets all languages granted via modifiers.
 */
export function getLanguageModifiers(sources: ModifierSource[]): string[] {
    const languages: string[] = [];

    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'language' && mod.value === true) {
                languages.push(mod.target);
            }
        }
    }

    return languages;
}

/**
 * Checks if a skill has expertise via modifiers.
 */
export function getExpertiseModifiers(skillName: string, sources: ModifierSource[]): boolean {
    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'expertise' &&
                mod.target.toLowerCase() === skillName.toLowerCase() &&
                mod.value === true) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Gets the total HP bonus per level from all sources.
 * Used for feats like Tough (+2/level) and traits like Dwarven Toughness (+1/level).
 * Target should be 'hp_per_level'.
 */
export function getHpPerLevelBonus(sources: ModifierSource[]): number {
    let totalBonus = 0;

    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'bonus' && mod.target.toLowerCase() === 'hp_per_level') {
                const val = typeof mod.value === 'number' ? mod.value : parseInt(String(mod.value), 10);
                if (!isNaN(val)) {
                    totalBonus += val;
                }
            }
        }
    }

    return totalBonus;
}

/**
 * Gets the total flat HP max bonus from all sources.
 * Used for features like Boon of Fortitude (+40 HP).
 * Target should be 'hp_max'.
 */
export function getHpMaxBonus(sources: ModifierSource[]): number {
    let totalBonus = 0;

    for (const source of sources) {
        for (const mod of source.modifiers) {
            if (mod.type === 'bonus' && mod.target.toLowerCase() === 'hp_max') {
                const val = typeof mod.value === 'number' ? mod.value : parseInt(String(mod.value), 10);
                if (!isNaN(val)) {
                    totalBonus += val;
                }
            }
        }
    }

    return totalBonus;
}

/**
 * Calculates the effective maximum HP including all modifiers.
 * Formula: baseHpMax + (level × hpPerLevelBonus) + flatHpMaxBonus
 */
export function calculateEffectiveMaxHp(
    baseHpMax: number,
    level: number,
    sources: ModifierSource[]
): number {
    const hpPerLevel = getHpPerLevelBonus(sources);
    const flatBonus = getHpMaxBonus(sources);

    return baseHpMax + (level * hpPerLevel) + flatBonus;
}

File: shared/src/commonMain/kotlin/org/dndcharacter/shared/mechanics/Modifiers.kt
@file:OptIn(kotlin.js.ExperimentalJsExport::class)

package org.dndcharacter.shared.mechanics

import kotlin.js.JsExport
import kotlin.js.JsName
import kotlin.math.min
import kotlin.math.max

@JsExport
data class RuleModifier(
    val type: String,
    val target: String,
    // Value can be number, boolean, or string in TS. 
    // In Kotlin, we need to handle this. For now, let's use String and parse it, 
    // or separate fields. Using String for interop simplicity, but specific accessors would be better.
    // simpler approach for KMP/JS interop: use separate nullable fields and a helper, 
    // OR just use String and cast/parse in logic.
    val valueStr: String, 
    val condition: String? = null,
    val max: Int? = null
) {
    // Helper to get number value
    fun getNumberValue(): Int = valueStr.toIntOrNull() ?: 0
    
    // Helper to get boolean value
    fun getBooleanValue(): Boolean = valueStr.lowercase() == "true"
}

@JsExport
data class ModifierSource(
    val id: String,
    val name: String,
    val modifiers: Array<RuleModifier>
)

@JsExport
object ModifierCalculator {

    /**
     * Calculates a stat value with modifiers applied.
     * 
     * Priority hierarchy (highest to lowest):
     * - override: Final authority. Highest override value wins.
     * - set: "At least this much." Highest set value wins. Applied if no override.
     * - bonus: Additive. All bonuses stack after set/override resolution.
     * 
     * @param baseValue The base stat value
     * @param statName The stat to calculate (e.g., "str", "ac", "speed")
     * @param sources All modifier sources to consider
     * @param conditionContext Optional condition context for filtering conditional modifiers
     */
    fun calculateStatWithModifiers(
        baseValue: Int, 
        statName: String, 
        sources: Array<ModifierSource>,
        conditionContext: ConditionEvaluator.ConditionContext? = null
    ): Int {
        var finalValue = baseValue
        val bonuses = mutableListOf<Int>()
        var setValues = mutableListOf<Int>() // Collect all "set" values
        var overrideValues = mutableListOf<Int>() // Collect all "override" values
        var minMaxCap: Int? = null
        val targetStat = statName.lowercase()

        for (source in sources) {
            for (mod in source.modifiers) {
                if (!mod.target.equals(targetStat, ignoreCase = true)) continue
                
                // Check conditions if context is provided
                if (conditionContext != null && !mod.condition.isNullOrEmpty()) {
                    val condition = org.dndcharacter.shared.model.ModifierCondition.fromKey(mod.condition)
                    if (!ConditionEvaluator.isConditionMet(condition, conditionContext)) {
                        continue // Condition not met, skip this modifier
                    }
                }
                // If no condition context but modifier has condition, skip (conservative)
                if (conditionContext == null && !mod.condition.isNullOrEmpty()) {
                    continue // TODO: Handle conditions when context not available
                }

                when (mod.type) {
                    "override" -> {
                        val v = mod.getNumberValue()
                        overrideValues.add(v)
                    }
                    "set" -> {
                        val v = mod.getNumberValue()
                        setValues.add(v)
                    }
                    "bonus", "ability_increase" -> {
                        val v = mod.getNumberValue()
                        bonuses.add(v)
                        
                        // Handle caps
                        if (mod.type == "ability_increase" && mod.max != null) {
                            minMaxCap = if (minMaxCap == null) mod.max else min(minMaxCap!!, mod.max)
                        }
                    }
                }
            }
        }

        // Priority resolution: override > set > base+bonuses
        // 1. If any override exists, use highest override (bonuses don't apply)
        if (overrideValues.isNotEmpty()) {
            finalValue = overrideValues.maxOrNull() ?: baseValue
        }
        // 2. If set exists, compare (base + bonuses) vs set - use higher
        else if (setValues.isNotEmpty()) {
            val basePlusBonuses = baseValue + bonuses.sum()
            val highestSet = setValues.maxOrNull() ?: baseValue
            finalValue = max(basePlusBonuses, highestSet)
        }
        // 3. Otherwise, apply bonuses to base
        else {
            finalValue = baseValue + bonuses.sum()
        }

        // Apply Max Cap (for ability_increase type modifiers)
        if (minMaxCap != null) {
            finalValue = min(finalValue, minMaxCap!!)
        }

        return finalValue
    }

    fun getProficiencyModifiers(statName: String, sources: Array<ModifierSource>): Boolean {
        val target = statName.lowercase()
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "skill_proficiency" && mod.target.equals(target, ignoreCase = true) && mod.getBooleanValue()) return true
                if (mod.type == "proficiency" && mod.target.equals(target, ignoreCase = true)) return true
            }
        }
        return false
    }

    fun getSavingThrowProficiency(ability: String, sources: Array<ModifierSource>): Boolean {
        val target = ability.lowercase()
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "saving_throw_proficiency" && 
                    mod.target.equals(target, ignoreCase = true) && 
                    mod.getBooleanValue()) return true
            }
        }
        return false
    }
    
    fun getArmorProficiency(armorType: String, sources: Array<ModifierSource>): Boolean {
        val target = armorType.lowercase()
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "armor_proficiency" && 
                    mod.target.equals(target, ignoreCase = true) && 
                    mod.getBooleanValue()) return true
            }
        }
        return false
    }
    
    fun getWeaponProficiency(weaponIdOrCategory: String, sources: Array<ModifierSource>): Boolean {
        val target = weaponIdOrCategory.lowercase()
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "weapon_proficiency" && mod.getBooleanValue()) {
                    if (mod.target.equals(target, ignoreCase = true)) return true
                }
            }
        }
        return false
    }
    
    fun getLanguageModifiers(sources: Array<ModifierSource>): Array<String> {
        val list = mutableListOf<String>()
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "language" && mod.getBooleanValue()) {
                    list.add(mod.target)
                }
            }
        }
        return list.toTypedArray()
    }
    
    fun getExpertiseModifiers(skillName: String, sources: Array<ModifierSource>): Boolean {
        val target = skillName.lowercase()
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "expertise" && 
                    mod.target.equals(target, ignoreCase = true) && 
                    mod.getBooleanValue()) return true
            }
        }
        return false
    }
    
    fun getHpPerLevelBonus(sources: Array<ModifierSource>): Int {
        var total = 0
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "bonus" && mod.target.equals("hp_per_level", ignoreCase = true)) {
                    total += mod.getNumberValue()
                }
            }
        }
        return total
    }
    
    fun getHpMaxBonus(sources: Array<ModifierSource>): Int {
        var total = 0
        for (source in sources) {
            for (mod in source.modifiers) {
                if (mod.type == "bonus" && mod.target.equals("hp_max", ignoreCase = true)) {
                    total += mod.getNumberValue()
                }
            }
        }
        return total
    }
    
    fun calculateEffectiveMaxHp(baseHpMax: Int, level: Int, sources: Array<ModifierSource>): Int {
        val hpPerLvl = getHpPerLevelBonus(sources)
        val flatBonus = getHpMaxBonus(sources)
        return baseHpMax + (level * hpPerLvl) + flatBonus
    }
    fun getSkillBonus(skillName: String, sources: Array<ModifierSource>): Int {
        val target = skillName.replace("_", " ").lowercase()
        var total = 0
        for (source in sources) {
            for (mod in source.modifiers) {
                // Check exact match or space/underscore replacement
                val modTarget = mod.target.replace("_", " ").lowercase()
                
                if (mod.type == "bonus" && modTarget == target) {
                    total += mod.getNumberValue()
                }
            }
        }
        return total
    }

    fun getSavingThrowBonus(ability: String, sources: Array<ModifierSource>): Int {
        val targetAbil = ability.lowercase()
        val targetSave = "${targetAbil}_save"
        var total = 0
        for (source in sources) {
            for (mod in source.modifiers) {
                val t = mod.target.lowercase()
                if (mod.type == "bonus") {
                    if (t == targetSave || t == "saving_throws" || t == "all_saving_throws") {
                        total += mod.getNumberValue()
                    }
                }
            }
        }
        return total
    }
}
